classes/migration/upgrade/v3_4_0/PreflightCheckMigration.php

```diff
@@ -32,752 +34,1148 @@ abstract protected function getContextTable(): string;
+     *
+     * @return array<string,string[]>
+     */
+    abstract protected function getEntityRelationships(): array;
+
+    /**
+     * Builds the array with all tables that require cleanup and their respective cleanup code
+     */
+    protected function buildOrphanedEntityProcessor(): void
+    {
+        $this->addTableProcessor('submissions', function (): int {
+            $affectedRows = 0;
+            // Depends directly on ~2 entities: context_id->context_table.context_id current_publication_id->publications.publication_id
+            // Custom field (not found in at least one of the softwares)
+            $affectedRows += $this->deleteRequiredReference('submissions', 'context_id', $this->getContextTable(), $this->getContextKeyField());
+
+            // Attempts to recover the field submissions.current_publication_id before discarding the entry
+            $rows = DB::table('submissions AS s')
+                ->leftJoin('publications AS p', 'p.publication_id', '=', 's.current_publication_id')
+                ->join(
+                    'publications AS last',
+                    fn(JoinClause $q) => $q->where(
+                        fn(Builder $q) => $q->from('publications AS p2')
+                            ->whereColumn('p2.submission_id', '=', 's.submission_id')
+                            ->orderByDesc('p2.publication_id')
+                            ->limit(1)
+                            ->select('p2.publication_id'),
+                        '=',
+                        DB::raw('last.publication_id')
+                    )
+                )
+                ->whereNull('p.publication_id')
+                ->pluck('s.submission_id', 'last.publication_id');
+            foreach ($rows as $publicationId => $submissionId) {
+                $this->_installer->log("The current publication ID ({$publicationId}) for the submission ID {$submissionId} is invalid, the publication ID {$publicationId} will replace it");
+                $affectedRows += DB::table('submissions')->where('submission_id', '=', $submissionId)->update(['current_publication_id' => $publicationId]);
             }
 
+            // The current_publication_id is nullable, but it's in fact required by the software, so we delete orphan entries instead of nulling them
+            $affectedRows += $this->deleteRequiredReference('submissions', 'current_publication_id', 'publications', 'publication_id');
+            return $affectedRows;
+        });
+
+        $this->addTableProcessor('submission_files', function (): int {
+            $affectedRows = 0;
+            // Depends directly on ~5 entities: file_id->files.file_id genre_id->genres.genre_id source_submission_file_id->submission_files.submission_file_id submission_id->submissions.submission_id uploader_user_id->users.user_id
+            $affectedRows += $this->deleteRequiredReference('submission_files', 'submission_id', 'submissions', 'submission_id');
```

Comment:
This causes a fatal error with the SciELO dataset on upgrade:
```
ERROR: Upgrade failed: DB: SQLSTATE[23000]: Integrity constraint violation: 1451 Cannot delete or update a parent row: a foreign key constraint fails (`scielo`.`submission_file_revisions`, CONSTRAINT `submission_file_revisions_submission_file_id_foreign` FOREIGN KEY (`submission_file_id`) REFERENCES `submission_files` (`submission_file_id`)) (SQL: delete from `submission_files` where `submission_id` in (68139, 68199, 68239, 68696, 68969, 69063, 69132, 69141, 69270, 69327, 69369, 69441, 69452, 69492, 69500, 69507, 69595, 69731, 69742, 69799, 69894, 69895, 70024, 70038, 70091, 70100, 70153, 70168, 70190, 70246, 70325, 70356, 70466, 70468, 70485, 70599, 70621, 70651, 70671, 70728, 71099, 71131, 71235, 71353, 71354, 71377, 71405, 71441, 71469, 71590, 71921, 71967, 71972, 72055, 72067, 72223, 72494, 72581, 72583, 72587, 72854, 72936, 72971, 72991, 73323, 73416, 73570, 73658, 73833, 73843, 73908, 74081, 74204, 74272, 74277, 74404, 74425, 74453, 74465, 74506, 74533, 74608, 74714, 74725, 74923, 74959, 74969, 74995, 75247, 75277, 75351, 75353, 75414, 75416, 75483, 75492, 75508, 75545, 75562, 75563, 75687, 75689, 75766, 75803, 75804, 75837, 75902, 75944, 75947, 75962, 75967, 75970, 75997, 76004, 76009, 76011, 76027, 76028, 76054, 76127, 76128, 76130, 76136, 76157, 76209, 76232, 76236, 76373, 76378, 76390, 76435, 76438, 76456, 76506, 76518, 76547, 76550, 76563, 76633, 76635, 76649, 76677, 76724, 76728, 76739, 76779, 76789, 76790, 76793, 76823, 76826, 76843, 76844, 76859, 76866, 76878, 76879, 76893, 76911, 76934, 76953, 76972, 76990, 76996, 77011, 77104, 77147, 77165, 77185, 77189, 77239, 77292, 77293, 77305, 77362, 77443, 77495, 77560, 77681, 77760, 77771, 77777, 77784, 77837, 78091, 78170, 78287, 78307, 78310, 78395, 78409, 78429, 78545, 78546, 78600, 78710, 78815, 78911, 78925, 78965, 79174, 79629, 79710, 79747, 79819, 79987, 80073, 80192, 80208, 80387, 80514, 80636, 80648, 80700, 80979, 81001, 81168, 81206, 81238, 81246, 81248, 81277, 81285, 81408, 81497, 81573, 81608, 81618, 81703, 81721, 81825, 81829, 81834, 81849, 81885, 81940, 82045, 82076, 82113, 82170, 82173, 82252, 82255, 82317, 82339, 82356, 82364, 82367, 82385, 82458, 82497, 82513, 82535, 82619, 82686, 82928, 83001, 83068, 83159, 83166, 83237, 83309, 83359, 83554, 83590, 83609, 83985, 83998, 84025, 84047, 84130, 84215, 84290, 84343, 84362, 84402, 84431, 84566, 84573, 84601, 84603, 84637, 84721, 84739, 84754, 84773, 84797, 84858, 84891, 84907, 84911, 84917, 84976, 84981, 84990, 84996, 85031, 85050, 85105, 85111, 85145, 85175, 85240, 85272, 85288, 85300, 85306, 85333, 85374, 85552, 85562, 85575, 85588, 85603, 85642, 85678, 85719, 85726, 85757, 85864, 85871, 85885, 86046, 86116, 86128, 86135, 86137, 86241, 86254, 86267, 86298, 86302, 86368, 86391, 86420, 86423, 86432, 86561, 86721, 86749, 86889, 86903, 87019, 87078, 87106, 87154, 87331, 87334, 87340, 87349, 87365, 87378, 87403, 87408, 87453, 87457, 87495, 87544, 87546, 87616, 87666, 87681, 87704, 87749, 87753, 87811, 87815, 87821, 87850, 87870, 87943, 87945, 87957, 87959, 88028, 88049, 88081, 88108, 88111, 88120, 88134, 88165, 88254, 88260, 88261, 88380, 88545, 88562, 88577, 88633, 88663, 88666, 88713, 88720, 88728, 88743, 88886, 88915, 88929, 88934, 88937, 88979, 88983, 89063, 89075, 89078, 89135, 89151, 89181, 89194, 89205, 89270, 89286, 89301, 89347, 89351, 89392, 89435, 89550, 89574, 89725, 90035, 90054, 90071, 90082, 90090, 90101, 90124, 90183, 90243, 90246, 90280, 90376, 90533, 90581, 90718, 90856, 90955, 90961, 90964, 90973, 91061, 91128, 91141, 91437, 91439, 91444, 91596, 91619, 91704, 91711, 91718, 91867, 91886, 91943, 91976, 92060, 92065, 92081, 92178, 92289, 92356, 92387, 92403, 92443, 92651, 92702, 92716, 92733, 92757, 92798, 92826, 92874, 92954, 93078, 93091, 93102, 93135, 93144, 93154, 93180, 93478, 93526, 93571, 93619, 93690, 93701, 93785, 93808, 93824, 93972, 94092, 94103, 94123, 94338, 94476, 94512, 94671, 94749, 94767, 94802, 94838, 94993, 95147, 95148, 95162, 95248, 95309, 95316, 95567, 95574, 95578, 95614, 95620, 95649, 95691, 95721, 95752, 95843, 95898, 96001, 96035, 96059, 96101, 96103, 96138, 96139, 96142, 96171, 96176, 96228, 96298, 96336, 96343, 96377, 96399, 96402, 96609, 96626, 96658, 96660, 96740, 96768, 96826, 96850, 96888, 96899, 96964, 97107, 97156, 97159, 97240, 97245, 97340, 97393, 97532, 97540, 97613, 97635, 97636, 97679, 97690, 97698, 97802, 97850, 97877, 97883, 97967, 97971, 97999, 98075, 98171, 98214, 98243, 98270, 98300, 98324, 98343, 98409, 98410, 98466, 98475, 98606, 98744, 98965, 99010, 99054, 99062, 99179, 99246, 99269, 99278, 99291, 99382, 99403, 99408, 99455, 99546, 99752, 99754, 99771, 99833, 99854, 99870, 99878, 99902, 99961, 100003, 100101, 100179, 100181, 100207, 100232, 100254, 100273, 100275, 100276, 100278, 100319, 100322, 100326, 100362, 100407, 100515, 100524, 100548, 100598, 100602, 100674, 100734, 100838, 100908, 100918, 100923, 100948, 101080, 101100, 101122, 101151, 101204, 101227, 101234, 101240, 101309, 101419, 101479, 101509, 101514, 101523, 101580, 101594, 101653, 101665, 101672, 101679, 101720, 101804, 101876, 101903, 101921, 101926, 101958, 102002, 102010, 102014, 102026, 102068, 102078, 102085, 102104, 102130, 102140, 102179, 102307, 102357, 102371, 102389, 102391, 102394, 102445, 102456, 102462, 102465, 102504, 102513, 102523, 102548, 102550, 102581, 102583, 102611, 102613, 102691, 102749, 102776, 102778, 102814, 102933, 102951, 102960, 102999, 103014, 103044, 103100, 103133, 103162, 103164, 103166, 103177, 103180, 103189, 103251, 103335, 103378, 103432, 103438, 103470, 103515, 103535, 103589, 103599, 103601, 103647, 103669, 103696, 103749, 103751, 103757, 103812, 103823, 103830, 103853, 103874, 103900, 103914, 103949, 104070, 104119, 104194, 104230, 104249, 104333, 104355, 104421, 104431, 104481, 104545, 104567, 104618, 104620, 104739, 104740, 104748, 104777, 104866, 104923, 104926, 104944, 104945, 104977, 104980, 104985, 104986, 104990, 104991, 104993, 105020, 105025, 105052, 105070, 105095, 105105, 105146, 105159, 105175, 105238, 105264, 105302, 105347, 105358, 105370, 105374, 105411, 105417, 105482, 105585, 105588, 105633, 105636, 105728, 105729, 105736, 105785, 105802, 105833, 105868, 105899, 105947, 105959, 106022, 106033, 106069, 106104, 106113, 106149, 106200, 106232, 106245, 106323, 106433, 106454, 106508, 106532, 106547, 106556, 106588, 106633, 106652, 106662, 106696, 106726, 106741, 106759, 106831, 106834, 106854, 106896, 106924, 106935, 106960, 107103, 107106, 107139, 107140, 107144, 107155, 107172, 107224, 107239, 107244, 107282, 107303, 107330, 107361, 107496, 107502, 107561, 107575, 107583, 107616, 107681, 107708, 107712, 107745, 107782, 107807, 107891, 107893, 107904, 107937, 107939, 107942, 107952, 107956, 107962, 107971, 107979, 107997, 108025, 108075, 108080, 108097, 108118, 108127, 108131, 108136, 108153, 108167, 108170, 108173, 108183, 108435, 108465, 108520, 108612, 108645, 108647, 108677, 108699, 108712, 108772, 108823, 108955, 108959, 108967, 109068, 109069, 109071, 109136, 109183, 109223, 109252, 109276, 109301, 109305, 109320, 109410, 109443, 109477, 109483, 109532, 109542, 109561, 109609, 109648, 109796, 109845, 109858, 109864, 109874, 110108, 110111, 110126, 110129, 110223, 110226, 110234, 110239, 110285, 110304, 110315, 110346, 110374, 110448, 110481, 110487, 110529, 110549, 110555, 110559, 110579, 110597, 110626, 110655, 110672, 110674, 110676, 110690, 110694, 110695, 110705, 110736, 110776, 110870, 110871, 110886, 110919, 110929, 110936, 110938, 110955, 110962, 110969, 111010, 111030, 111073, 111075, 111078, 111107, 111155, 111163, 111167, 111188, 111246, 111265, 111267, 111306, 111310, 111372, 111388, 111483, 111488, 111500, 111510, 111557, 111586, 111670, 111699, 111708, 111709, 111734, 111780, 111846, 111847, 111896, 111936, 111951, 111971, 111986, 112000) or `submission_id` is null)
```
I suspect it's because of the condition noted here:

https://github.com/pkp/pkp-lib/pull/8970/files#diff-6bd4382888007c21acb6bf540ce1eb2ba123dcc507ab2fb20edc3f97e36181c8R62

Reply:
I get a similar problem with review_round_files:
```
ERROR: Upgrade failed: DB: SQLSTATE[23000]: Integrity constraint violation: 1451 Cannot delete or update a parent row: a foreign key constraint fails (`scielo`.`review_round_files`, CONSTRAINT `review_round_files_submission_file_id_foreign` FOREIGN KEY (`submission_file_id`) REFERENCES `submission_files` (`submission_file_id`)) (SQL: delete from `submission_files` where `submission_id` in (68139, 68199, 68239, 68696, 68969, 69063, 69132, 69141, 69270, 69327, 69369, 69441, 69452, 69492, 69500, 69507, 69595, 69731, 69742, 69799, 69894, 69895, 70024, 70038, 70091, 70100, 70153, 70168, 70190, 70246, 70325, 70356, 70466, 70468, 70485, 70599, 70621, 70651, 70671, 70728, 71099, 71131, 71235, 71353, 71354, 71377, 71405, 71441, 71469, 71590, 71921, 71967, 71972, 72055, 72067, 72223, 72494, 72581, 72583, 72587, 72854, 72936, 72971, 72991, 73323, 73416, 73570, 73658, 73833, 73843, 73908, 74081, 74204, 74272, 74277, 74404, 74425, 74453, 74465, 74506, 74533, 74608, 74714, 74725, 74923, 74959, 74969, 74995, 75247, 75277, 75351, 75353, 75414, 75416, 75483, 75492, 75508, 75545, 75562, 75563, 75687, 75689, 75766, 75803, 75804, 75837, 75902, 75944, 75947, 75962, 75967, 75970, 75997, 76004, 76009, 76011, 76027, 76028, 76054, 76127, 76128, 76130, 76136, 76157, 76209, 76232, 76236, 76373, 76378, 76390, 76435, 76438, 76456, 76506, 76518, 76547, 76550, 76563, 76633, 76635, 76649, 76677, 76724, 76728, 76739, 76779, 76789, 76790, 76793, 76823, 76826, 76843, 76844, 76859, 76866, 76878, 76879, 76893, 76911, 76934, 76953, 76972, 76990, 76996, 77011, 77104, 77147, 77165, 77185, 77189, 77239, 77292, 77293, 77305, 77362, 77443, 77495, 77560, 77681, 77760, 77771, 77777, 77784, 77837, 78091, 78170, 78287, 78307, 78310, 78395, 78409, 78429, 78545, 78546, 78600, 78710, 78815, 78911, 78925, 78965, 79174, 79629, 79710, 79747, 79819, 79987, 80073, 80192, 80208, 80387, 80514, 80636, 80648, 80700, 80979, 81001, 81168, 81206, 81238, 81246, 81248, 81277, 81285, 81408, 81497, 81573, 81608, 81618, 81703, 81721, 81825, 81829, 81834, 81849, 81885, 81940, 82045, 82076, 82113, 82170, 82173, 82252, 82255, 82317, 82339, 82356, 82364, 82367, 82385, 82458, 82497, 82513, 82535, 82619, 82686, 82928, 83001, 83068, 83159, 83166, 83237, 83309, 83359, 83554, 83590, 83609, 83985, 83998, 84025, 84047, 84130, 84215, 84290, 84343, 84362, 84402, 84431, 84566, 84573, 84601, 84603, 84637, 84721, 84739, 84754, 84773, 84797, 84858, 84891, 84907, 84911, 84917, 84976, 84981, 84990, 84996, 85031, 85050, 85105, 85111, 85145, 85175, 85240, 85272, 85288, 85300, 85306, 85333, 85374, 85552, 85562, 85575, 85588, 85603, 85642, 85678, 85719, 85726, 85757, 85864, 85871, 85885, 86046, 86116, 86128, 86135, 86137, 86241, 86254, 86267, 86298, 86302, 86368, 86391, 86420, 86423, 86432, 86561, 86721, 86749, 86889, 86903, 87019, 87078, 87106, 87154, 87331, 87334, 87340, 87349, 87365, 87378, 87403, 87408, 87453, 87457, 87495, 87544, 87546, 87616, 87666, 87681, 87704, 87749, 87753, 87811, 87815, 87821, 87850, 87870, 87943, 87945, 87957, 87959, 88028, 88049, 88081, 88108, 88111, 88120, 88134, 88165, 88254, 88260, 88261, 88380, 88545, 88562, 88577, 88633, 88663, 88666, 88713, 88720, 88728, 88743, 88886, 88915, 88929, 88934, 88937, 88979, 88983, 89063, 89075, 89078, 89135, 89151, 89181, 89194, 89205, 89270, 89286, 89301, 89347, 89351, 89392, 89435, 89550, 89574, 89725, 90035, 90054, 90071, 90082, 90090, 90101, 90124, 90183, 90243, 90246, 90280, 90376, 90533, 90581, 90718, 90856, 90955, 90961, 90964, 90973, 91061, 91128, 91141, 91437, 91439, 91444, 91596, 91619, 91704, 91711, 91718, 91867, 91886, 91943, 91976, 92060, 92065, 92081, 92178, 92289, 92356, 92387, 92403, 92443, 92651, 92702, 92716, 92733, 92757, 92798, 92826, 92874, 92954, 93078, 93091, 93102, 93135, 93144, 93154, 93180, 93478, 93526, 93571, 93619, 93690, 93701, 93785, 93808, 93824, 93972, 94092, 94103, 94123, 94338, 94476, 94512, 94671, 94749, 94767, 94802, 94838, 94993, 95147, 95148, 95162, 95248, 95309, 95316, 95567, 95574, 95578, 95614, 95620, 95649, 95691, 95721, 95752, 95843, 95898, 96001, 96035, 96059, 96101, 96103, 96138, 96139, 96142, 96171, 96176, 96228, 96298, 96336, 96343, 96377, 96399, 96402, 96609, 96626, 96658, 96660, 96740, 96768, 96826, 96850, 96888, 96899, 96964, 97107, 97156, 97159, 97240, 97245, 97340, 97393, 97532, 97540, 97613, 97635, 97636, 97679, 97690, 97698, 97802, 97850, 97877, 97883, 97967, 97971, 97999, 98075, 98171, 98214, 98243, 98270, 98300, 98324, 98343, 98409, 98410, 98466, 98475, 98606, 98744, 98965, 99010, 99054, 99062, 99179, 99246, 99269, 99278, 99291, 99382, 99403, 99408, 99455, 99546, 99752, 99754, 99771, 99833, 99854, 99870, 99878, 99902, 99961, 100003, 100101, 100179, 100181, 100207, 100232, 100254, 100273, 100275, 100276, 100278, 100319, 100322, 100326, 100362, 100407, 100515, 100524, 100548, 100598, 100602, 100674, 100734, 100838, 100908, 100918, 100923, 100948, 101080, 101100, 101122, 101151, 101204, 101227, 101234, 101240, 101309, 101419, 101479, 101509, 101514, 101523, 101580, 101594, 101653, 101665, 101672, 101679, 101720, 101804, 101876, 101903, 101921, 101926, 101958, 102002, 102010, 102014, 102026, 102068, 102078, 102085, 102104, 102130, 102140, 102179, 102307, 102357, 102371, 102389, 102391, 102394, 102445, 102456, 102462, 102465, 102504, 102513, 102523, 102548, 102550, 102581, 102583, 102611, 102613, 102691, 102749, 102776, 102778, 102814, 102933, 102951, 102960, 102999, 103014, 103044, 103100, 103133, 103162, 103164, 103166, 103177, 103180, 103189, 103251, 103335, 103378, 103432, 103438, 103470, 103515, 103535, 103589, 103599, 103601, 103647, 103669, 103696, 103749, 103751, 103757, 103812, 103823, 103830, 103853, 103874, 103900, 103914, 103949, 104070, 104119, 104194, 104230, 104249, 104333, 104355, 104421, 104431, 104481, 104545, 104567, 104618, 104620, 104739, 104740, 104748, 104777, 104866, 104923, 104926, 104944, 104945, 104977, 104980, 104985, 104986, 104990, 104991, 104993, 105020, 105025, 105052, 105070, 105095, 105105, 105146, 105159, 105175, 105238, 105264, 105302, 105347, 105358, 105370, 105374, 105411, 105417, 105482, 105585, 105588, 105633, 105636, 105728, 105729, 105736, 105785, 105802, 105833, 105868, 105899, 105947, 105959, 106022, 106033, 106069, 106104, 106113, 106149, 106200, 106232, 106245, 106323, 106433, 106454, 106508, 106532, 106547, 106556, 106588, 106633, 106652, 106662, 106696, 106726, 106741, 106759, 106831, 106834, 106854, 106896, 106924, 106935, 106960, 107103, 107106, 107139, 107140, 107144, 107155, 107172, 107224, 107239, 107244, 107282, 107303, 107330, 107361, 107496, 107502, 107561, 107575, 107583, 107616, 107681, 107708, 107712, 107745, 107782, 107807, 107891, 107893, 107904, 107937, 107939, 107942, 107952, 107956, 107962, 107971, 107979, 107997, 108025, 108075, 108080, 108097, 108118, 108127, 108131, 108136, 108153, 108167, 108170, 108173, 108183, 108435, 108465, 108520, 108612, 108645, 108647, 108677, 108699, 108712, 108772, 108823, 108955, 108959, 108967, 109068, 109069, 109071, 109136, 109183, 109223, 109252, 109276, 109301, 109305, 109320, 109410, 109443, 109477, 109483, 109532, 109542, 109561, 109609, 109648, 109796, 109845, 109858, 109864, 109874, 110108, 110111, 110126, 110129, 110223, 110226, 110234, 110239, 110285, 110304, 110315, 110346, 110374, 110448, 110481, 110487, 110529, 110549, 110555, 110559, 110579, 110597, 110626, 110655, 110672, 110674, 110676, 110690, 110694, 110695, 110705, 110736, 110776, 110870, 110871, 110886, 110919, 110929, 110936, 110938, 110955, 110962, 110969, 111010, 111030, 111073, 111075, 111078, 111107, 111155, 111163, 111167, 111188, 111246, 111265, 111267, 111306, 111310, 111372, 111388, 111483, 111488, 111500, 111510, 111557, 111586, 111670, 111699, 111708, 111709, 111734, 111780, 111846, 111847, 111896, 111936, 111951, 111971, 111986, 112000) or `submission_id` is null)
```

Reply:
...and review_files:
```
ERROR: Upgrade failed: DB: SQLSTATE[23000]: Integrity constraint violation: 1451 Cannot delete or update a parent row: a foreign key constraint fails (`scielo`.`review_files`, CONSTRAINT `review_files_submission_file_id_foreign` FOREIGN KEY (`submission_file_id`) REFERENCES `submission_files` (`submission_file_id`)) (SQL: delete from `submission_files` where `submission_id` in (170577, 170608, 170645, 170668, 170682, 170691, 170754, 170767, 170774, 170790, 170838, 170971, 170985, 171059, 171062, 171089, 171093, 171139, 171146, 171150, 171205, 171230, 171252, 171265, 171306, 171319, 171406, 171409, 171462, 171502, 171509, 171522, 171529, 171555, 171559, 171560, 171614, 171625, 171673, 171676, 171707, 171715, 171720, 171725, 171727, 171739, 171764, 171781, 171897, 171985, 171998, 172035, 172045, 172078, 172123, 172159, 172162, 172164, 172168, 172214, 172238, 172275, 172310, 172332, 172391, 172398, 172402, 172409, 172421, 172425, 172450, 172452, 172487, 172491, 172513, 172570, 172578, 172624, 172707, 172744, 172884, 172896, 172919, 172926, 172932, 172953, 172958, 172971, 172983, 172995, 173020, 173024, 173072, 173104, 173110, 173139, 173150, 173176, 173242, 173265, 173323, 173347, 173368, 173380, 173381, 173435, 173453, 173454, 173468, 173469, 173472, 173474, 173482, 173490, 173503, 173528, 173535, 173551, 173555, 173556, 173563, 173565, 173579, 173597, 173599, 173611, 173643, 173644, 173672, 173705, 173717, 173738, 173747, 173802, 173803, 173831, 173834, 173842, 173856, 173864, 173891, 173949, 173964, 173985, 174018, 174152, 174188, 174205, 174214, 174218, 174273, 174295, 174407, 174409, 174417, 174506, 174523, 174541, 174572, 174605, 174663, 174667, 174670, 174680, 174778, 174782, 174821, 174872, 174877, 174911, 174934, 174937, 174938, 174966, 175083, 175129, 175180, 175181, 175207, 175212, 175215, 175219, 175258, 175319, 175321, 175323, 175347, 175372, 175384, 175406, 175436, 175472, 175476, 175558, 175584, 175596, 175664, 175674, 175700, 175741, 175752, 175756, 175771, 175777, 175781, 175798, 175800, 175818, 175863, 175878, 175900, 175917, 175956, 175970, 176002, 176009, 176036, 176040, 176091, 176107, 176314, 176333, 176334, 176358, 176387, 176441, 176488, 176538, 176559, 176582, 176614, 176678, 176710, 176734, 176788, 176845, 176855, 176862, 176942, 176947, 176963, 176977, 177031, 177032, 177061, 177074, 177124, 177127, 177203, 177262, 177263, 177351, 177367, 177380, 177385, 177408, 177497, 177564, 177567, 177602, 177619, 177641, 177652, 177690, 177698, 177727, 177738, 177752, 177779, 177785, 177812, 177824, 177859, 177896, 177905, 177915, 177932, 177999, 178137, 178143, 178145, 178150, 178193, 178216, 178246, 178273, 178274, 178285, 178290, 178299, 178303, 178338, 178340, 178349, 178372, 178385, 178390, 178392, 178393, 178397, 178408, 178419, 178420, 178426, 178440, 178451, 178452, 178455, 178461, 178482, 178492, 178568, 178571, 178616, 178640, 178660, 178668, 178724, 178765, 178833, 178891, 178907, 178945, 179003, 179115, 179160, 179182, 179207, 179232, 179240, 179266, 179307, 179324, 179362, 179414, 179436, 179460, 179488, 179495, 179517, 179532, 179564, 179565, 179588, 179643, 179646, 179674, 179680, 179686, 179732, 179788, 179790, 179811, 179824, 179870, 179883, 179972, 179979, 180046, 180155, 180163, 180196, 180207, 180218, 180238, 180256, 180314, 180411, 180464, 180467, 180468, 180482, 180547, 180568, 180570, 180589, 180733, 180745, 180772, 180826, 180847, 180854, 180866, 180904, 180917, 180919, 180920, 180930, 180943, 180944, 180945, 180995, 181111, 181117, 181131, 181156, 181258, 181290, 181333, 181337, 181417, 181451, 181481, 181573, 181587, 181600, 181643, 181649, 181663, 181682, 181700, 181702, 181723, 181732, 181743, 181766, 181827, 181836, 181843, 181893, 181981, 182029, 182096, 182212, 182213, 182246, 182282, 182320, 182334, 182356, 182411, 182440, 182451, 182456, 182475, 182527, 182561, 182578, 182582, 182614, 182619, 182657, 182678, 182692, 182723, 182745, 182792, 182900, 182925, 182966, 183008, 183041, 183085, 183213, 183282, 183305, 183423, 183426, 183509, 183511, 183513, 183541, 183549, 183697, 183760, 183762, 183763, 183857, 183862, 183868, 183899, 183915, 183928, 183970, 183979, 183997, 184002, 184015, 184021, 184025, 184061, 184080, 184091, 184117, 184139, 184165, 184179, 184189, 184205, 184240, 184244, 184247, 184258, 184342, 184380, 184400, 184422, 184438, 184439, 184506, 184515, 184523, 184530, 184589, 184600, 184604, 184614, 184615, 184651, 184661, 184667, 184675, 184686, 184696, 184709, 184734, 184735, 184740, 184747, 184751, 184835, 184874, 184878, 184880, 184921, 184937, 184950, 184970, 185002, 185010, 185045, 185068, 185138, 185156, 185186, 185191, 185212, 185372, 185418, 185445, 185487, 185510, 185514, 185519, 185579, 185600, 185624, 185634, 185673, 185733, 185759, 185775, 185809, 185831, 185862, 185866, 185891, 185892, 185921, 185960, 185975, 185984, 185990, 186036, 186065, 186103, 186110, 186112, 186131, 186176, 186201, 186209, 186224, 186245, 186269, 186274, 186279, 186281, 186395, 186418, 186501, 186506, 186525, 186544, 186555, 186587, 186588, 186594, 186702, 186708, 186723, 186724, 186759, 186771, 186833, 186834, 186947, 186949, 187006, 187013, 187034, 187037, 187107, 187131, 187137, 187166, 187176, 187201, 187206, 187209, 187231, 187288, 187291, 187328, 187336, 187365, 187377, 187392, 187545, 187574, 187580, 187592, 187597, 187663, 187692, 187714, 187765, 187791, 187811, 187854, 187919, 187985, 188023, 188095, 188172, 188182, 188201, 188204, 188222, 188225, 188245, 188251, 188279, 188283, 188306, 188313, 188318, 188327, 188400, 188463, 188486, 188497, 188522, 188594, 188657, 188725, 188731, 188735, 188739, 188773, 188787, 188829, 188833, 188866, 188872, 188882, 188885, 188888, 188957, 188960, 188974, 189036, 189066, 189069, 189096, 189104, 189124, 189143, 189170, 189188, 189211, 189262, 189265, 189284, 189296, 189386, 189398, 189440, 189512, 189678, 189779, 189920, 189948, 189972, 190015, 190034, 190136, 190206, 190265, 190289, 190307, 190415, 190509, 190644, 190713, 190732, 190772, 190806, 190827, 190844, 190847, 190869, 190874, 190875, 190897, 190904, 190916, 190921, 190928, 190967, 190979, 190982, 191010, 191059, 191126, 191161, 191181, 191184, 191187, 191198, 191250, 191252, 191358, 191464, 191486, 191526, 191535, 191547, 191633, 191635, 191689, 191696, 191732, 191734, 191752, 191753, 191885, 191918, 191925, 191982, 192048, 192088, 192132, 192198, 192202, 192212, 192284, 192349, 192368, 192388, 192540, 192596, 192644, 192651, 192656, 192657, 192688, 192704, 192717, 192751, 192755, 192781, 192788, 192789, 192791, 192793, 192908, 192921, 192952, 192965, 192995, 193004, 193005, 193025, 193045, 193050, 193084, 193111, 193133, 193142, 193143, 193166, 193167, 193195, 193205, 193224, 193292, 193352, 193394, 193426, 193451, 193495, 193521, 193535, 193632, 193710, 193744, 193772, 193810, 193843, 193883, 193893, 193897, 193902, 193924, 193939, 194035, 194037, 194038, 194102, 194141, 194227, 194275, 194487, 194568, 194730, 194817, 194871, 194892, 195036, 195099, 195288, 195423, 195564, 195621, 195639, 195663, 195696, 195738, 196059, 196203, 196215, 196323, 196350, 196419, 196512, 197268, 197271, 197334, 197382, 197586, 197646, 197679, 197790, 197847, 198066, 198120, 198447, 198504, 198516, 198678, 198852, 198888, 198915, 198933, 199008, 199140, 199467, 199734, 199914, 200148, 200178, 200181, 200349, 200409, 200496, 200502, 200574, 200637, 200706, 200808, 200919, 200991, 200997, 201021, 201030, 201036, 201261, 201291, 201492, 201504, 201537, 201909, 201990, 202059, 202131, 202170, 202188, 202242, 202383, 202536, 202545, 202617, 202692, 202725, 202728, 202803, 202911, 203040, 203064, 203100, 203178, 203283, 203313, 203583, 203658, 203685, 203742, 203844, 203874, 204237, 204324, 204426, 204450, 204462, 204501, 204519, 204726, 204831, 204882, 204897, 205158, 205164, 205221, 205227, 205242, 205422, 205494, 205533, 205548, 205551, 205566, 205593, 205632, 205668, 205683, 205704, 205716, 205725, 205743, 205809, 205860, 205962, 206247, 206343, 206379, 206499, 206628, 206706, 206718, 206733, 206823, 206931, 207066, 207138, 207291, 207357, 207411, 207507, 207579, 207594, 207624, 207717, 207759, 207825, 207960, 208008, 208032, 208212, 208317, 208377, 208518, 208545, 209028, 209160, 209163, 209250, 209319, 209622, 209625, 209853, 209871, 209922, 209931, 209958, 209988, 209991, 209994, 210075, 210087) or `submission_id` is null)
```

Reply:
It makes sense to fail. Dropping the problematic FKs should be placed here instead (that's why I think the "preflight" should only do checks/fast validations, otherwise it ends up handling too many details) :)

Reply:
FYI I did what I've suggested above, and added code to clear the FKs (I didn't move, just duplicated code from the original place).
The previous migration wasn't cleaning these entities, that's why it didn't fail.